### PR TITLE
Fixes issues 16271, 18825, 8065, 5050 - Allow function literals return by reference

### PR DESCRIPTION
**Pull request by BorisCarvajal**
respective change in spec dlang/dlang.org#2618

### CODE REVIEW

**Comment by thewilsonator** [comment]
Please take a look at the buildkite ocean failures
```
./src/ocean/transition.d(42): Error: basic type expected, not `static`
--
  | ./src/ocean/transition.d(42): Error: found `static` when expecting `)`
  | ./src/ocean/transition.d(42): Error: semicolon expected following auto declaration, not `immutable`
  | ./src/ocean/transition.d(42): Error: declaration expected, not `)`
  | ./src/ocean/meta/types/Qualifiers.d(73): Error: basic type expected, not `static`
  | ./src/ocean/meta/types/Qualifiers.d(73): Error: found `static` when expecting `)`
  | ./src/ocean/meta/types/Qualifiers.d(73): Error: found `immutable` when expecting `)`
  | ./src/ocean/meta/types/Qualifiers.d(73): Error: found `)` when expecting `;`
  | ./src/ocean/meta/types/Qualifiers.d(73): Error: found `)` instead of statement
  | ./src/ocean/meta/types/Qualifiers.d(169): Error: basic type expected, not `static`
  | ./src/ocean/meta/types/Qualifiers.d(169): Error: found `static` when expecting `)`
  | ./src/ocean/meta/types/Qualifiers.d(169): Error: found `immutable` when expecting `)`
  | ./src/ocean/meta/types/Qualifiers.d(169): Error: no identifier for declarator `U`
  | ./src/ocean/meta/types/Qualifiers.d(169): Error: declaration expected, not `)`
  | ./src/ocean/meta/types/Qualifiers.d(173): Error: declaration expected, not `else`
  | ./src/ocean/meta/types/Qualifiers.d(176): Error: unrecognized declaration
  | ./src/ocean/core/Test.d(230): Error: semicolon expected following function declaration
  | ./src/ocean/core/Test.d(232): Error: declaration expected, not `if`
```

**Comment by JinShil** [comment]
> Please take a look at the buildkite ocean failures

This may have been caused by [Ocean's recent 5.0.0 release](https://github.com/sociomantic-tsunami/ocean/releases).

**Comment by BorisCarvajal** [comment]
I checked and the error is the same with stable.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by ZombineDev** [comment]
Thanks for working on this @BorisCarvajal!

**Comment by RazvanN7** [comment]
Could you please add the text "Fix Issue" followed by the issue numbers in a commit so that the dlang bot automatically links the issues with this PR?

**Comment by thewilsonator** [comment]
Specifically the commit message should contain "Fixes issues 16271, 18825, 8065, 5050", i.e. make it the same as the PR title.

**Comment by BorisCarvajal** [comment]
done

**Comment by jacob-carlborg** [comment]
* It looks like all examples are using `=>`. Perhaps add some examples with using curly braces. 
* What about declaring a variable of a delegate type that returns ref, is that already possible?

**Comment by BorisCarvajal** [comment]
* I just tested the refness of the returns, the syntax is fairly proved already.
I was going to put some checks with those but I thought was overkill, I'll add a few anyways.
* Yes, you can do:
```
int x;
ref int dg() { return x; }
auto d = &dg; // or alias d = dg;
```

File: src/dmd/parse.d [hunk 1]
```diff
@@ -8038,6 +8058,22 @@ final class Parser(AST) : Lexer
...
+                    Token* tk = peekPastParen(peek(&token));
+                    if (skipAttributes(tk, &tk) && (tk.value == TOK.goesTo || tk.value == TOK.leftCurly))
+                    {
+                        // ref (arguments) => expression
+                        // ref (arguments) { statements... }
+                        goto case_delegate;
+                    }
+                }
+                nextToken();
+                error("found `%s` when expecting function literal following %s", token.toChars(), "`ref`".ptr);
```
Comment (ZombineDev) [inline src/dmd/parse.d]: Can you add a test for this error message as well?

**Review by PetarKirov — APPROVED** [review]

File: src/dmd/parse.d (hunk 1 above)
Comment (BorisCarvajal) [reply src/dmd/parse.d]: done. I'm not that confident about the message though.

Comment (ZombineDev) [reply src/dmd/parse.d]: I think it's ok. I don't have better suggestions at the moment.

Comment (Basile-z) [inline src/dmd/parse.d]: You can directly put the second variadic argument  in the format specifier.

Comment (BorisCarvajal) [reply src/dmd/parse.d]: Indeed, I think I read the error a few lines above and copied the format.
What a waste of typing.

**Comment by jacob-carlborg** [comment]
Make sure the commits are squashed when merging.

**Comment by thewilsonator** [comment]
@ZombineDev I had the spec PR on 72h -> merge, I don't think we need to wait for both.

**Review by thewilsonator — APPROVED** [review]

**Comment by wilzbach** [comment]
@BorisCarvajal as this is definitely a new feature that DMD users should be made aware of (e.g. so that they can update their D grammar/parsers), it would be great if you could follow up with a small changelog entry:

https://github.com/dlang/dmd/tree/master/changelog

Thanks!

This also will help to expose your work to the wider D community.

**Comment by andralex** [comment]
Great work. Thanks @BorisCarvajal !

**Pull request merged**